rfd/0057-automatic-aws-server-discovery.md

```diff
@@ -0,0 +1,391 @@
+```
+
+In order to run the new SSM document the AWS user will need IAM permissions to run
+SSM commands[3] for example:
+
+```js
+{
+    "Statement": [
+        {
+            "Action": "ssm:SendCommand",
+            "Effect": "Allow",
+            "Resource": [
+                # Allow running commands on all us-west-2 instances
+                "arn:aws:ssm:us-west-2:*:instance/*",
+                 # Allows running the installTeleport docuemnt on the allowed instances
+                "arn:aws:ssm:us-east-2:aws-account-ID:document/installTeleport"
+            ]
+        },
+		// "CreateDocument" and "GetDocument" permissions are required
+		// to automatically create the document
+        {
+            "Action": "ssm:CreateDocument",
+            "Effect": "Allow",
+            "Resource": [ "*" ]
+        },
+		{
+            "Action": "ssm:GetDocument",
+            "Effect": "Allow",
+            "Resource": [ "*" ]
+        }		
+    ]
+}
+```
+
+The machines being discovered will need to allow recieving `ec2messages` in
+order to recieve the SSM commands:
+
+```js 
+{
+	"Statement": [
+		{
+			"Action": "ec2messages:GetMessages"
+			"Effect": "Allow"
+		}
+	]
+}
+```
+
+On AWS, Amazon Linux and Ubuntu LTS (16.04, 18.04, 20.04) come with the SSM agent
+preinstalled[4].
+
+In order to allow nodes to create tokens for the purposes of sending invites to EC2
+instances a new system role will be added -- `RoleNodeDiscovery`, that will have
+permissions to create tokens.
+
+Each EC2 instance that is to be discovered will also require that they have an IAM
+role attached, in order to be able to send and recieve messages for the SSM agent.
+
+Example:
+
+```json
+{
+    "Statement": [
+        {
+            "Action": "ec2messages:*",
+            "Effect": "Allow",
+            "Resource": [
+                # Allow running commands on all us-west-2 instances
+                "*"
+            ]
+        }
+    ]
+}
+```
+
+
+## teleport.yaml generation
+
+The `teleport node configure` subcommand will be used to generate a
+new /etc/teleport.yaml file:
+```sh
+teleport node configure
+    --auth-server=auth-server.example.com [auth server that is being connected to]
+    --token="$1" # passed via parameter from SSM document
+    --labels=teleport.dev/origin=cloud
+	--node-name=$accountID-$instanceID
+```
+This will create generate a file with the following contents:
+
+```yaml
+teleport: 
+  nodename: "$accountID-$instanceID"
+  auth_servers:
+    - "auth-server.example.com:3025"
+  join_params:
+    token_name: token
+ssh_service:
+  enabled: yes
+  labels:
+    teleport.dev/origin: "cloud"
+```
+
+## UX
+
+### User has 1 account to discover servers on
+
+#### Teleport config
+
+Discovery server:
+```yaml
+teleport:
+  ...
+auth_service:
+  enabled: "yes"
+ssh_service:
+  enabled: "yes"
+  aws:
+  - types: ["ec2"]
+    regions: ["us-west-1"]
+    tags:
+      "teleport": "yes" # aws tags to match
```

Comment:
How do I specify what SSM command and script to run? Would it just run one script for everything? Can we add a script doc name as an optional parameter here? We should create a default one, but it can't be singleton

Reply:
Yes, it works like you describe. We create a default SSM document but you can optionally specify a different one to use. There's a configuration example somewhere above:

```yaml
  aws:
   - types: ["ec2"]
     regions: ["eu-central-1"]
     tags:
       "teleport": "yes"
     ssm:
       document: "TeleportDiscoveryInstaller" # default value
```

@lxea Please correct me if I'm wrong.